### PR TITLE
ENH: add get_best_config_with_material_priority and tests

### LINKED ISSUE #47: Change the filter calculation method (material-based prioritization)
Currently
----------
The "best" configuration is the one closest to the desired attenuation, regardless of filter material.

The ESD notes (which is confirmed by @mseaberg) that this is not the approach that should be taken, as the silicon filters should be protected by the diamond filters.

ESD
----

According to the ESD (LCLSII-3.5-ES-0267-R1, page 8):
> Diamond filters shall always be used when the silicon filters are in the beam (pre-filtering) per Figure 4.

Page 11:
> Diamond filter(s) shall always be used to provide attenuation (pre-filtering) when silicon filters are used. The required attenuation and diamond filter thicknesses for pre-filtering the silicon filters in the 6 – 25 keV range is shown in Figure 4.

Page 15:
> When operating in the Cu-HXR mode, the MPS shall ensure that:
> [...]
> * The appropriate diamond pre-filter is inserted prior to inserting a silicon filter,

<details>
<summary>LCLSII-3.5-ES-0267-R1 / Figure 4</summary>
image
</details>

Scientist response
-------------------
Matt noted:
> [...] the intention of the diamond filters [is for] protecting the silicon. [...] until they are needed to reduce transmission further, the silicon filters would remain out. This would mean that any time a silicon filter is inserted, all of the diamond filters should already be in. This is only really an issue at low photon energies which we mostly won’t be using this run (aside from possibly during AD machine studies), but it would be good to get this functioning the right way.

and:
> This is not super urgent since [...] this mainly matters at low photon energies (and/or high pulse energies, which we’re currently not getting).

Path forward
--------------
There is existing code that could be used as a basis for using all of the diamond filters before using the silicon filters, written by @mseaberg. We should look into integrating it well before it's actually needed on the beamline.

I'm hoping it can be done in a somewhat generic way for re-usability between the systems, support stuck filters, etc.

cc @slacAWallace

### LINKED ISSUE #58: Verify filter calculations in test suite
* Add some quick test cases
* Use test cases generated from `lcls_beamline_toolbox`-based example

<details>

```python
    def _get_thickness_both(self, transmission, E):
        """
        Calculate silicon thickness in microns to achieve the desired
        transmission
        """
        diamond_thickness = self.diamond.get_thickness(transmission,
                                                       E * 1e3) * 1e6
        if diamond_thickness > self.max_diamond:
            diamond_thickness = self.max_diamond

        Td = self.diamond.transmission(thickness=diamond_thickness * 1e-6)

        energy = self.diamond.energy
        Td = np.interp(E * 1e3, energy, Td)

        Ts = transmission / Td

        silicon_thickness = self.silicon.get_thickness(Ts, E * 1e3) * 1e6

        return diamond_thickness, silicon_thickness

    def _get_thickness(self, transmission, E):
        """
        Calculate silicon thickness in microns to achieve the desired
        transmission
        """
        thickness = self.silicon.get_thickness(transmission, E * 1e3) * 1e6

        return thickness

    def blade_insertions(self, transmission, E):
        d_thickness, s_thickness = self._get_thickness_both(transmission, E)

        # figure out binary representation of transmission with powers of 2
        d_multiplier = d_thickness / 10
        d_multiplier = int(d_multiplier)
        # print(np.binary_repr(int(d_multiplier), width=8))

        d_blade_insertions = np.binary_repr(d_multiplier, width=8)

        # figure out binary representation of transmission with powers of 2
        s_multiplier = s_thickness / 20
        s_multiplier = int(s_multiplier)
        # print(s_multiplier)
        # print(np.binary_repr(int(s_multiplier), width=10))

        s_blade_insertions = np.binary_repr(s_multiplier, width=10)

        return d_blade_insertions, s_blade_insertions

```
</details>

### PR DESCRIPTION
**Pull request by klauer**
Context
---------
See linked issues for context:
Closes #58 
Closes #47 

Steps
------
- [x] Add `get_best_config_with_material_priority`
- [x] Add basic tests
- [x] Integrate with the IOC (removing floor/ceil support for now)

Tests
-----
Tests are passing locally, but failing on Travis. This will be resolved at a later date.

Future
-------
(Separate PR/issue) I think in its current form, `find_configs` won't work for the SXR ladder-style variant and will require further thought.

### CODE REVIEW

**Comment by klauer** [comment]
Have a preliminary working material prioritization function that matches up with those expected from Matt's beamline config prototype.

~I don't much like its API, so that may get refactored prior to merging, along with some different tests.~

File: solid_attenuator/ioc_lfe_at2l0_calc/calculator.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 import itertools
 import pathlib
 import typing
+from typing import Tuple
```
Comment (ZLLentz) [inline solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:20]: Stylistically, how do you pick when to use `import typing` vs `from typing import ...`?

Comment (klauer) [reply solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:20]: It's usually `typing.X` until I start using `X` a few times - or there's a scenario where `typing.X` doesn't fit on one line but `X` does...

File: solid_attenuator/ioc_lfe_at2l0_calc/calculator.py [hunk 2]
```diff
@@ -167,6 +187,73 @@ def get_best_config(all_transmissions: typing.List[float],
...
+def get_best_config_with_material_priority(
+        materials: typing.List[str],
+        transmissions: typing.List[float],
+        material_order: typing.List[str],
+        t_des: float,
+        ) -> Config:
+    """
+    Inserting filters based on the material order provided, get the best
+    possible filter configuration.
+
```
Comment (ZLLentz) [inline solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:200]: This should have a short blurb explaining that the materials with lower priority will not be used at all until the materials with higher priority are all inserted

Comment (klauer) [reply solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:200]: Noted this in 5186917

Comment (ZLLentz) [reply solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:200]: I said this because it was in the attached issues, but I don't see where in the code this is accomplished... How does this work? All the tests seem to pass...

Comment (klauer) [reply solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:200]: Very good question - I didn't explain this nearly as well as I should have. Let me try here before reformatting and attempting to put into code. How's this sound?

The original algorithm
--------------------
The generic "best configuration" algorithm does its best to choose the right filters to target the requested transmission.
It does this by enumerating all possible filter configurations, doing a matrix multiplication to determine the transmission values for all of those configurations. Figuring out the best configuration is then a matter of matching the closest transmission value (i.e., `argmin(abs(all_transmissions - target_transmission))`.  This makes no assumptions as to how the filters are laid out and is as generic as possible.

Now, if all filters of every material type are used with the above algorithm, some silicon filters may be inserted prior to diamond filters. This is not good, as the silicon filters may be damaged. We need a new algorithm to get around that.

This PR
--------
The proposed algorithm breaks up the problem into 2 calls to the original algorithm, once per material.

Using the example of AT2L0, with 8 diamond filters and 10 silicon filters:

(1) Apply the original algorithm _only_ for the diamond filters. Due to how the filters are physically configured, the diamond filters can get very close to the required transmission up to a certain value (`t_all_diamond_inserted` to `1.0`). This works because the filter thicknesses were carefully specified: each filter is 2x the previous filter in thickness. 

As an aside, this means you could treat each filter as binary value, spanning the range  `t_all_diamond_inserted <= t_des <= 1.0` in 2^num_filters = 2^8 steps (i.e., `(1 - t_all_diamond_inserted) / 2 ** 8`, which is `< 0.4%` per step)

Let's call the result from step (1) a transmission value of` transmission_1`.

(2) Now,
(a) For desired transmission within `t_all_diamond_inserted <= t_des <= 1.0`, the algorithm will have pretty much gotten the requested transmission as close as needed.
(b) For desired transmission `0 <= t_des < t_all_diamond_inserted`, the algorithm will have already selected all diamond filters. There's more left to do in the second round as we can still get closer to `t_des`, when calculating which silicon filters to insert.

(3) As normalized transmission values are multiplicative (50% of 50% transmission would be 25%), our second round tries to find the right configuration of silicon filters for `t_des / transmission_1`. 
For values in the range noted by (2a), `t_des / transmission1` will be close to 1 - so no filters will be inserted.
For values in the range noted by (2b), `t_des / transmission1` will likely be something actionable.

(4) Assembling the two configurations from (1) and (3) gives a final configuration, specifying all filter states.

Comment (ZLLentz) [reply solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:200]: Ok, so it stops at step `2a` and avoids putting in `Si` in these cases by forcing the calculation to slightly overshoot the attenuation (undershoot the transmission) such that adding additional `Si` filters will only make us get further away from the goal attenuation? And that is why I was confused by the code which doesn't have any additional stopping parameter here? I think this was my misunderstanding point.

Comment (klauer) [reply solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:200]: Yes, the choice of 'ceiling' in step 2a overshoots the transmission, ensuring that no silicon filters are inserted unless necessary. That's a good thing to clarify in the docs as well - thanks for looking through this.

Comment (klauer) [reply solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:200]: _Actually_, the actually should have been `floor` to safely insert the silicon filters, if for some reason diamond filters stop working (i.e., fail in the out position).  Brute forcing energy ranges and transmission values confirms the above.

This does change the test suite cases and makes the results not match up 1-to-1 with what Matt had determined previously. I'll have to think about how to rework them in a sensible way; perhaps by adding the brute-force verification of configurations based on energy, desired transmission, and non-working filters to the test suite.

Comment (ZLLentz) [reply solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:200]: That's an annoying edge case for sure

**Comment by klauer** [comment]
Some interesting things during testing. The following plots are based on the filter material priority algorithm:

<details>
<summary>Plots</summary>

The `x` marker indicates the normalized transmission value when all _diamond_ filters are inserted at the annotated energy.

1. Assuming all filters are working:
image
As transmission is a function of material, thickness, and the photon energy, we see the algorithm failing to match what the user requests at lower energies and above `t_all_diamond`. This is where the silicon filters may not be inserted to compensate.

2. Assuming just the 1280µm diamond filter is stuck out:
image

3. Assuming the two 10µm and 20µm diamond filters are stuck out:
image

The above show the piecewise behavior you'd expect from discrete filters, and show the calculations doing roughly doing what we ask. While it was a useful exercise for me, I'm not sure there's much to gain from it.  

Regardless, the above is now in the test suite, along with checks to ensure the algorithm doesn't insert silicon filters when it shouldn't.
</details>

Separately, I've reintroduced floor/ceiling and explicit checks to be extra-certain that all diamond filters are marked for insertion prior to silicon filters. I think the tests are reasonably exhaustive and straightforward now, so 🤞 this should be good to go.

File: solid_attenuator/ioc_lfe_at2l0_calc/calculator.py (hunk 2 above)
Comment (klauer) [reply solid_attenuator/ioc_lfe_at2l0_calc/calculator.py:200]: OK, all of the above is captured in the docs now, slightly modified from the above due to some algorithm changes I made today.

**Comment by klauer** [comment]
Merging before the break, before I forget that this was in (what I think) an OK state.
